moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -385,4 +385,101 @@ std::shared_ptr<BaseConstraint> createOMPLConstraint(const moveit::core::RobotMo
     return nullptr;
   }
 }
-}  // namespace ompl_interface
\ No newline at end of file
+
+/******************************************
+ * Toolpath Constraint
+ * ****************************************/
+ToolPathConstraint::ToolPathConstraint(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group,
+                                       unsigned int num_dofs, unsigned int steps)
+  : BaseConstraint(robot_model, group, num_dofs), num_steps_(steps)
+{
+  NearestNeighbors<geometry_msgs::msg::Pose>::DistanceFunction func =
+      std::bind(&poseDistance, std::placeholders::_1, std::placeholders::_2);
+
+  pose_nn_.setDistanceFunction(func);
+}
+
+void ToolPathConstraint::setPath(const moveit_msgs::msg::CartesianTrajectory& msg)
+{
+  // TODO: Should this be done in the constructor? What happens when we want to accept another msg?
+
+  // Clear previous path
+  pose_nn_.clear();
+
+  // Interpolate along waypoints and insert into pose_nn_;
+  const auto dt = 1.0 / static_cast<double>(num_steps_);
+  for (unsigned int i = 0; i < msg.points.size() - 1; i++)
+  {
+    // Insert waypoints
+    pose_nn_.add(msg.points.at(i).point.pose);
+
+    // Insert interpolated poses
+    auto step = dt;
+    for (unsigned int j = 0; j < num_steps_ - 1; j++)
```

Comment:
it might make more sense to have a variable that indicates the max spacing between waypoints and use that to determine how many interpolated poses you need between any pair of waypoints.